### DISCUSSION TITLE
I have indexed my projects but search turns up empty

### DISCUSSION THREAD

**Discussion by daniel-dlds** [oracle/opengrok#4675] [General]
Hi,

I am running opengrok 1.13.23 on jdk 23 on raspbian. I have three projects.
1 in C, 1 in Cpp and another in Python.
I can browse the files through the web interface.
But searching for anything gives me an empty result.

In the tomcat log I can see : 
org.opengrok.suggest.SuggesterProjectData.lookup No WFST for field full in <path to data/suggester>

What do I have to do in order to fix this ?

Thanks,
Regards

**Comment by vladak** [comment]
The log message means that the suggester did not find its data. Even if the suggester was not initialized, it would not impact searching. Does your OpenGrok data root include indexes for the 3 projects ? How was indexer run and are there any relevant messages in the indexer logs ? What happens if you restart Tomcat ?

**Comment by daniel-dlds** [comment]
Hi,
Thanks for your reply.

I restarted tomcat and the log of the deployment of opengrok webapp is below : 

`15-Oct-2024 14:10:48.122 INFO [main] org.apache.catalina.startup.HostConfig.deployWAR Deploying web application archive [/portabledrive/javaapps/apache-tomcat-10.1.26/webapps/source.war]
15-Oct-2024 14:10:48.142 WARNING [main] org.apache.tomcat.util.digester.SetPropertiesRule.begin Match [Context] failed to set property [antiJARLocking] to [true]
15-Oct-2024 14:10:52.080 INFO [main] org.apache.jasper.servlet.TldScanner.scanJars At least one JAR was scanned for TLDs yet contained no TLDs. Enable debug logging for this logger for a complete list of JARs that were scanned but no TLDs were found in them. Skipping unneeded JARs during scanning can improve startup time and JSP compilation time.
15-Oct-2024 14:10:52.303 INFO [main] org.opengrok.indexer.Metrics.<clinit> configuring PrometheusRegistry
15-Oct-2024 14:10:52.453 INFO [main] org.opengrok.web.WebappListener.contextInitialized Starting webapp with version 1.13.23 (fc478df1d21db874df32b93375390ae6af19110b)
15-Oct-2024 14:10:52.456 INFO [main] org.opengrok.indexer.configuration.Configuration.read Reading configuration from '/var/opengrok/etc/configuration.xml'
15-Oct-2024 14:10:53.036 INFO [main] org.opengrok.indexer.util.Statistics.logIt Done invalidating repositories (3 valid, 3 working) (took 453 ms)
15-Oct-2024 14:10:53.047 INFO [main] org.opengrok.web.WebappListener.contextInitialized running inside Apache Tomcat/10.1.26
15-Oct-2024 14:10:53.053 WARNING [main] org.opengrok.indexer.framework.PluginFramework.reload Plugin directory '/portabledrive/javaapps/opengrok-1.13.23/data/../plugins' not found or not readable: All requests allowed.
15-Oct-2024 14:10:53.060 INFO [main] org.opengrok.indexer.configuration.Configuration.read Reading configuration from '/var/opengrok/etc/configuration.xml'
15-Oct-2024 14:10:53.073 INFO [OpenGrok-index-check-41] org.opengrok.indexer.index.IndexCheck.checkDirWithLogging Checking index in '/portabledrive/javaapps/opengrok-1.13.23/data/index/Boobs' (mode VERSION)
15-Oct-2024 14:10:53.073 INFO [OpenGrok-index-check-40] org.opengrok.indexer.index.IndexCheck.checkDirWithLogging Checking index in '/portabledrive/javaapps/opengrok-1.13.23/data/index/ctoolboxgeneral' (mode VERSION)
15-Oct-2024 14:10:53.094 WARNING [OpenGrok-index-check-41] org.apache.lucene.store.MMapDirectory.lookupProvider You are running with Java 22 or later. To make full use of MMapDirectory, please update Apache Lucene.
15-Oct-2024 14:10:53.346 INFO [OpenGrok-index-check-41] org.opengrok.indexer.index.IndexCheck.checkDirWithLogging Index check for directory '/portabledrive/javaapps/opengrok-1.13.23/data/index/Boobs' passed
15-Oct-2024 14:10:53.346 INFO [OpenGrok-index-check-40] org.opengrok.indexer.index.IndexCheck.checkDirWithLogging Index check for directory '/portabledrive/javaapps/opengrok-1.13.23/data/index/ctoolboxgeneral' passed
15-Oct-2024 14:10:53.347 INFO [OpenGrok-index-check-40] org.opengrok.indexer.index.IndexCheck.checkDirWithLogging Checking index in '/portabledrive/javaapps/opengrok-1.13.23/data/index/mainWebApp' (mode VERSION)
15-Oct-2024 14:10:53.349 INFO [OpenGrok-index-check-40] org.opengrok.indexer.index.IndexCheck.checkDirWithLogging Index check for directory '/portabledrive/javaapps/opengrok-1.13.23/data/index/mainWebApp' passed
15-Oct-2024 14:10:53.874 INFO [Thread-9] org.opengrok.suggest.Suggester.init Initializing suggester for [mainWebApp, ctoolboxgeneral, Boobs]
15-Oct-2024 14:10:53.988 WARNING [Thread-9] org.opengrok.suggest.SuggesterProjectData.initFields Fields [defs, path, hist, refs, type, full] will be ignored because they were not found in index directory 'MMapDirectory@/portabledrive/javaapps/opengrok-1.13.23/data/index/mainWebApp lockFactory=org.apache.lucene.store.NativeFSLockFactory@31c7336f'
15-Oct-2024 14:10:54.011 WARNING [Thread-9] org.opengrok.suggest.SuggesterProjectData.initFields Fields [defs, path, hist, refs, type, full] will be ignored because they were not found in index directory 'MMapDirectory@/portabledrive/javaapps/opengrok-1.13.23/data/index/ctoolboxgeneral lockFactory=org.apache.lucene.store.NativeFSLockFactory@31c7336f'
15-Oct-2024 14:10:54.031 WARNING [Thread-9] org.opengrok.suggest.SuggesterProjectData.initFields Fields [defs, path, hist, refs, type, full] will be ignored because they were not found in index directory 'MMapDirectory@/portabledrive/javaapps/opengrok-1.13.23/data/index/Boobs lockFactory=org.apache.lucene.store.NativeFSLockFactory@31c7336f'
15-Oct-2024 14:10:54.080 INFO [Thread-9] org.opengrok.suggest.Suggester.init Suggesters for [mainWebApp, ctoolboxgeneral, Boobs] were initialized (took 0 seconds)
15-Oct-2024 14:10:54.105 INFO [Thread-9] org.opengrok.web.api.v1.suggester.provider.service.impl.SuggesterServiceImpl.scheduleRebuild Scheduling suggester rebuild in PT9H49M5.915816204S
15-Oct-2024 14:10:55.084 INFO [main] org.hibernate.validator.internal.util.Version.<clinit> HV000001: Hibernate Validator 8.0.1.Final
15-Oct-2024 14:10:55.491 INFO [main] org.apache.catalina.startup.HostConfig.deployWAR Deployment of web application archive [/portabledrive/javaapps/apache-tomcat-10.1.26/webapps/source.war] has finished in [7,369] ms`

I also did a tail -f on the opengrok specific log and in the startup process it didnt write anything.

In the opengrok installation under data/index I have a subdirectory for each one of the projects with some files having lucene in their name. I also have a directory data/suggester with a subdirectory for each project. But the only file in those is named version.txt.
Should these contain more files ?

I have a script that runs the indexer on the tomcat server : 

java -jar /portabledrive/javaapps/opengrok-1.13.23/lib/opengrok.jar -s /portabledrive/javaapps/opengrok-1.13.23/src/ -d /portabledrive/javaapps/opengrok-1.13.23/data/ -R /var/opengrok/etc/configuration.xml -v

The output of this script is : 

`Oct 15, 2024 2:23:00 PM org.opengrok.indexer.configuration.Configuration read
INFO: Reading configuration from '/var/opengrok/etc/configuration.xml'
Oct 15, 2024 2:23:00 PM org.opengrok.indexer.index.Indexer parseOptions
INFO: Indexer options: [-s, /portabledrive/javaapps/opengrok-1.13.23/src/, -d, /portabledrive/javaapps/opengrok-1.13.23/data/, -R, /var/opengrok/etc/configuration.xml, -v]
Oct 15, 2024 2:23:00 PM org.opengrok.indexer.configuration.Configuration checkConfiguration
INFO: History based reindex is on, however history is off. History has to be enabled for history based reindex.
Logging filehandler pattern: %h/java%u.log
SLF4J(W): No SLF4J providers were found.
SLF4J(W): Defaulting to no-operation (NOP) logger implementation
SLF4J(W): See https://www.slf4j.org/codes.html#noProviders for further details.
Oct 15, 2024 2:23:01 PM org.opengrok.indexer.util.Statistics logIt
INFO: Done invalidating repositories (3 valid, 3 working) (took 429 ms)
Oct 15, 2024 2:23:01 PM org.opengrok.indexer.index.Indexer main
INFO: Indexer version 1.13.23 (fc478df1d21db874df32b93375390ae6af19110b) running on Java 23
Oct 15, 2024 2:23:01 PM org.opengrok.indexer.analysis.Ctags lambda$run$0
WARNING: Error from ctags: ctags: Warning: Unknown language "terraform:+.tf.Tf.TF" in "langmap" option
Oct 15, 2024 2:23:01 PM org.opengrok.indexer.analysis.Ctags lambda$run$0
WARNING: Error from ctags: ctags: Warning: Unknown language "terraform:+.tfvars.Tfvars.TFVARS" in "langmap" option
Oct 15, 2024 2:23:01 PM org.opengrok.indexer.configuration.RuntimeEnvironment validateUniversalCtags
INFO: Using ctags: Universal Ctags 5.9.0, Copyright (C) 2015 Universal Ctags Team
Universal Ctags is derived from Exuberant Ctags.
Exuberant Ctags 5.8, Copyright (C) 1996-2009 Darren Hiebert
  Compiled: Sep  3 2021, 18:12:18
  URL: https://ctags.io/
  Optional compiled features: +wildcards, +regex, +gnulib_regex, +iconv, +option-directory, +xpath, +json, +interactive, +sandbox, +yaml, +packcc, +optscript
Oct 15, 2024 2:23:01 PM org.opengrok.indexer.index.Indexer prepareIndexer
INFO: Generating history cache for all repositories ...
Oct 15, 2024 2:23:01 PM org.opengrok.indexer.history.HistoryGuru createHistoryCacheReal
INFO: Creating history cache for 3 repositories
Oct 15, 2024 2:23:01 PM org.opengrok.indexer.history.HistoryGuru createHistoryCache
INFO: Creating history cache for {dir='/portabledrive/javaapps/opengrok-1.13.23/src/Boobs',type=git,historyCache=on,renamed=false,merge=true,annotationCache=off}
Oct 15, 2024 2:23:01 PM org.opengrok.indexer.history.HistoryGuru createHistoryCache
INFO: Creating history cache for {dir='/portabledrive/javaapps/opengrok-1.13.23/src/ctoolboxgeneral',type=git,historyCache=on,renamed=false,merge=true,annotationCache=off}
Oct 15, 2024 2:23:01 PM org.opengrok.indexer.history.HistoryGuru createHistoryCache
INFO: Creating history cache for {dir='/portabledrive/javaapps/opengrok-1.13.23/src/mainWebApp',type=git,historyCache=on,renamed=false,merge=true,annotationCache=off}
Oct 15, 2024 2:23:05 PM org.opengrok.indexer.util.Statistics logIt
INFO: Done history cache for {dir='/portabledrive/javaapps/opengrok-1.13.23/src/ctoolboxgeneral',type=git,historyCache=on,renamed=false,merge=true,annotationCache=off} (took 4.313 seconds)
Oct 15, 2024 2:23:06 PM org.opengrok.indexer.util.Statistics logIt
INFO: Done history cache for {dir='/portabledrive/javaapps/opengrok-1.13.23/src/mainWebApp',type=git,historyCache=on,renamed=false,merge=true,annotationCache=off} (took 4.482 seconds)
Oct 15, 2024 2:23:06 PM org.opengrok.indexer.util.Statistics logIt
INFO: Done history cache for {dir='/portabledrive/javaapps/opengrok-1.13.23/src/Boobs',type=git,historyCache=on,renamed=false,merge=true,annotationCache=off} (took 5.40 seconds)
Oct 15, 2024 2:23:06 PM org.opengrok.indexer.util.Statistics logIt
INFO: Done history cache for all repositories (took 5.61 seconds)
Oct 15, 2024 2:23:06 PM org.opengrok.indexer.index.Indexer prepareIndexer
INFO: Done generating history cache
Oct 15, 2024 2:23:06 PM org.opengrok.indexer.util.Statistics logIt
INFO: Done invalidating repositories (3 valid, 3 working) (took 27 ms)
Oct 15, 2024 2:23:06 PM org.opengrok.indexer.index.Indexer doIndexerExecution
INFO: Starting indexing
Oct 15, 2024 2:23:06 PM org.apache.lucene.store.MMapDirectory lookupProvider
WARNING: You are running with Java 22 or later. To make full use of MMapDirectory, please update Apache Lucene.
Oct 15, 2024 2:23:06 PM org.opengrok.indexer.index.Indexer doIndexerExecution
INFO: Waiting for the executors to finish
Oct 15, 2024 2:23:07 PM org.opengrok.indexer.index.IndexDatabase getIndexDownArgs
INFO: Starting file collection using history traversal for directory '/ctoolboxgeneral'
Oct 15, 2024 2:23:07 PM org.opengrok.indexer.index.IndexDatabase getIndexDownArgs
INFO: Starting file collection using history traversal for directory '/Boobs'
Oct 15, 2024 2:23:07 PM org.opengrok.indexer.index.IndexDatabase getIndexDownArgs
INFO: Starting file collection using history traversal for directory '/mainWebApp'
Oct 15, 2024 2:23:07 PM org.opengrok.indexer.util.Statistics logIt
INFO: Done file collection for directory '/mainWebApp' (took 455 ms)
Oct 15, 2024 2:23:07 PM org.opengrok.indexer.index.IndexDatabase update
INFO: Starting indexing of directory '/mainWebApp'
Oct 15, 2024 2:23:07 PM org.opengrok.indexer.analysis.Ctags lambda$run$0
WARNING: Error from ctags: ctags: Warning: Unknown language "terraform:+.tf.Tf.TF" in "langmap" option
Oct 15, 2024 2:23:07 PM org.opengrok.indexer.analysis.Ctags lambda$run$0
WARNING: Error from ctags: ctags: Warning: Unknown language "terraform:+.tfvars.Tfvars.TFVARS" in "langmap" option
Oct 15, 2024 2:23:07 PM org.opengrok.indexer.analysis.Ctags lambda$run$0
WARNING: Error from ctags: ctags: Warning: Unknown language "terraform:+.tf.Tf.TF" in "langmap" option
Oct 15, 2024 2:23:07 PM org.opengrok.indexer.analysis.Ctags lambda$run$0
WARNING: Error from ctags: ctags: Warning: Unknown language "terraform:+.tfvars.Tfvars.TFVARS" in "langmap" option
Oct 15, 2024 2:23:07 PM org.opengrok.indexer.analysis.Ctags lambda$run$0
WARNING: Error from ctags: ctags: Warning: Unknown language "terraform:+.tf.Tf.TF" in "langmap" option
Oct 15, 2024 2:23:07 PM org.opengrok.indexer.analysis.Ctags lambda$run$0
WARNING: Error from ctags: ctags: Warning: Unknown language "terraform:+.tfvars.Tfvars.TFVARS" in "langmap" option
Oct 15, 2024 2:23:07 PM org.opengrok.indexer.util.Statistics logIt
INFO: Done file collection for directory '/ctoolboxgeneral' (took 845 ms)
Oct 15, 2024 2:23:07 PM org.opengrok.indexer.index.IndexDatabase update
INFO: Starting indexing of directory '/ctoolboxgeneral'
Oct 15, 2024 2:23:08 PM org.opengrok.indexer.util.Statistics logIt
INFO: Done file collection for directory '/Boobs' (took 1.398 seconds)
Oct 15, 2024 2:23:08 PM org.opengrok.indexer.index.IndexDatabase update
INFO: Starting indexing of directory '/Boobs'
Oct 15, 2024 2:23:08 PM org.opengrok.indexer.analysis.Ctags lambda$run$0
WARNING: Error from ctags: ctags: Warning: ignoring null tag in /portabledrive/javaapps/opengrok-1.13.23/src/mainWebApp/shampoos/static/jquery-3.2.1.js(line: 6540)
Oct 15, 2024 2:23:08 PM org.opengrok.indexer.analysis.Ctags lambda$run$0
WARNING: Error from ctags: ctags: Warning: ignoring null tag in /portabledrive/javaapps/opengrok-1.13.23/src/mainWebApp/shampoos/static/jquery-3.2.1.js(line: 8090)
Oct 15, 2024 2:23:08 PM org.opengrok.indexer.analysis.Ctags lambda$run$0
WARNING: Error from ctags: ctags: Warning: Unknown language "terraform:+.tf.Tf.TF" in "langmap" option
Oct 15, 2024 2:23:08 PM org.opengrok.indexer.analysis.Ctags lambda$run$0
WARNING: Error from ctags: ctags: Warning: Unknown language "terraform:+.tfvars.Tfvars.TFVARS" in "langmap" option
Oct 15, 2024 2:23:09 PM org.opengrok.indexer.util.Statistics logIt
INFO: Done indexing of directory '/mainWebApp' (took 2.342 seconds)
Oct 15, 2024 2:23:10 PM org.opengrok.indexer.util.Statistics logIt
INFO: Done indexing of directory '/ctoolboxgeneral' (took 3.21 seconds)
Oct 15, 2024 2:23:13 PM org.opengrok.indexer.util.Statistics logIt
INFO: Done indexing of directory '/Boobs' (took 5.86 seconds)
Oct 15, 2024 2:23:14 PM org.opengrok.indexer.util.Statistics logIt
INFO: Done indexing data of all repositories (took 8.58 seconds)
Oct 15, 2024 2:23:14 PM org.opengrok.indexer.util.Statistics logIt
INFO: Indexer finished (took 14.342 seconds)`

I don't know what component is responsible for the search function : the indexer or the suggester ?

Thanks,
Regards

**Reply by vladak** [reply]
The webapp is responsible for the search. Suggester is just an extra feature to provide live suggestions as you type; it is not necessary and in fact can be completely disabled without altering the classic search.

The indexer creates the indexes which the webapp accesses in order to provide search. In your case, based on the indexer logs and the fact that the projects are visible in the webapp, the indexing was successful, however it does not seem to be able to access the index for some reason.

Based on this message:
```
WARNING [OpenGrok-index-check-41] org.apache.lucene.store.MMapDirectory.lookupProvider You are running with Java 22 or later. To make full use of MMapDirectory, please update Apache Lucene.
```
my guess is that Java 23 is little bit too new for the deployment. The current language level for OpenGrok is Java 11.

**Comment by daniel-dlds** [comment]
Hi,

I installed Jdk-11 and started OpenGrok with that. Now on application deployment I'm getting errors in the tomcat log like : 

16-Oct-2024 12:01:23.677 WARNING [main] org.apache.tomcat.util.scan.StandardJarScanner.processURLs Failed to scan [file:/p
ortabledrive/javaapps/apache-tomcat-10.1.26/webapps/source/WEB-INF/lib/jakarta.el-api-5.0.1.jar] from classloader hierarch
y

I have downloaded tomcat 9, 10 and 11. None of them has that jar or any other version of it. I also looked at the opengrok distribution and it's not there. As I suspect that it should belong to tomcat as this jar in particular is the implementation of the expression language.

Which version of tomcat can I use ? Or do I have to add all the jars that opengrok complains manually ?

EDIT : I have found for example the file jakarta.el-api-5.0.0.jar in the WEB-INF lib of the exploded WAR file inside tomcat.
Why is it asking for 5.0.1 ?

EDIT2 : In the opengrok lib directory of the distribution is the 5.0.1 version of the library. Why does the correct library exists outside the war file, and the wrong is in WEB-INF lib of the war file ?

Thanks,
Regards

**Reply by vladak** [reply]
These warnings are innocuous - see #3390.

As for the Tomcat, version 10 is the minimum.

**Comment by daniel-dlds** [comment]
I copied all the jars on the opengrok distribution to the exploded WEB-INF/lib and the exceptions are gone.
As for the initial problem, it also has gone away. The search function now works.

Thank you for your assistance,
Regards
